### ISSUE TITLE
Lint request: `avoid_private_access_on_open_classes`

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/linter#4975]
In Dart, private members _cannot_ be implemented outside of their declaring library, and as such, using `implements A` where `A` has a private member that is accessed outside of it's class can easily create an invalid class state; _other_ types assume subtype `A` will always have a private member that it _cannot_ have by definition.

An example described by @jakemac53:

```dart
// a.dart
class A {
  int get _private => 0;
}

void foo(A a) => print(a._private);
```

```dart
// b.dart
import 'a.dart';

// OK, no hints/lints/compile issues.
class B implements A {
  // Even if this was commented-back in it would not help.
  // int get _private => 0;
}
```

I'd like to suggest a _lint_ (that hopefully we add to recommended sets as it matures), which I'm naming `avoid_private_access_on_open_classes` (but the name is not important to me, only the semantics), it would flag private members accessed _outside_ of their declaring class or sub-classes declared in the declaring library:

```dart
class A {
  int get _private => 0;

  @override
  // OK
  String toString() => 'A: $_private';
}

class B extends A {
  @override
  // OK
  String toString() => 'B: $_private';
}

void foo(A a) {
  // LINT: avoid_private_access_on_open_classes
  print(a._private);
}

class C {
  final A a;
  C(this.a);
  
  @override
  // LINT: avoid_private_access_on_open_classes
  String toString() => 'C: ${a._private}';
}

final class D {
  int get _private => 0;
}

void d(D d) {
  // OK, class is final.
  print(d._private);
}

base class E {
  int get _private => 0;
}

void e(E e) {
  // OK, class is base.
  print(e._private);
}
```

The error message/quick-fixes could suggest:

1. Making `A` `final`
2. Making `A` `base`
3. Making `_private` public
4. Silencing the lint with an `// ignore` if the developer knows this type is not truly API public

---

User-land issue: https://github.com/flutter/flutter/issues/148692.

**Comment by srawlins** [comment]
It took me a few readings to get the motivation, but I think I get it. The key that I was missing is that this access is _OK_ on _final_ classes, because we _know_ the runtime type (the exhaustive list anyhow). But for an _open_ class, anyone could have claimed to implement it, and you _know_ that they didn't really implement it.

**Comment by matanlurey** [comment]
Yup you got it!

**Comment by FMorschel** [comment]
Just to show the importance of this lint, there are probably some other Flutter APIs which are being used this way. 

As I had already filed https://github.com/flutter/flutter/issues/148033. There are probably more "controllers" and similar classes that currently have that issue.

For these cases, another solution would be creating a wrapper class or something similar that would implement that private functionality. That class could potentially be private and that would be fine, I _think_.

**Comment by matanlurey** [comment]
> > I don't see any potential for runtime exceptions?
> 
> You can extend `A` outside this library afaik, and will get the runtime error on calls to `toString()`.

@jakemac53 I think that's fine, right? The A#toString() call accesses A#_private, which is fine?

**Comment by jakemac53** [comment]
> @jakemac53 I think that's fine, right? The A#toString() call accesses A#_private, which is fine?

It isn't fine, because A is abstract and has no implementation of `_private`, but B is still allowed to extend it from outside the library, so you end up with a runtime stack trace in `A#toString()`:

```
Unhandled exception:
NoSuchMethodError: Class 'B' has no instance getter '_private'.
Receiver: Instance of 'B'
Tried calling: _private
#0      B._private (package:hello_world/b.dart:8:7)
#1      A.toString (package:hello_world/a.dart:6:26)
```

**Comment by FMorschel** [comment]
Not showing the error but the actual code running:

<details>

<summary> bug.dart </summary>

```dart
import 'other.dart';

void main() {
  final a = A();
  final b = B();
  final c = C();
  print(a.toString());
  print(b.toString());
  print(c.toString());
}

class A {
  int get _private => 0;
  @override
  String toString() => '$runtimeType: $_private';
}
```

</details>

<details>

<summary> other.dart </summary>

```dart
import 'bug.dart';

class B implements A {}

class C extends A {}
```

</details>

Output:

```
A: 0
Instance of 'B'
C: 0

Exited.
```

**Comment by bwilkerson** [comment]
> Is it possible to flag every unsafe private member invocation, with a reasonably small number of false positives?

I would postulate that compile-time errors should have zero false positives. It's fine if the language forbids a construct that _could_ be given a runtime semantic, but if the construct is allowed by the language then we can't report it as an error.

Warnings need to be similarly free of false positives, though we do currently have a few violations of the policy.

We could, however, create a lint for this if we think we'd enable it in the core lint set (and it doesn't require whole-program analysis). But for a lint we don't need to flag _every_ violation in order for it to be reasonable, so there might be a tradeoff between completeness and the number of false positives.

**Comment by eernstg** [comment]
> I would postulate that compile-time errors should have zero false positives
> ...
> if the construct is allowed by the language then we can't report it as an error.

I thought lints would freely report situations that aren't compile-time errors, that's basically all they can do?

In any case, we can under-approximate and over-approximate (obtaining a guarantee that no non-problems are reported, perhaps failing to detect some actual problems, respectively obtaining a guarantee that no problems remain unreported, perhaps reporting some situations that aren't actual problems). We just need to choose an approach which is in line with existing policies.

**Comment by bwilkerson** [comment]
> > if the construct is allowed by the language then we can't report it as an error.
>
> I thought lints would freely report situations that aren't compile-time errors, that's basically all they can do?

Sorry for the confusion. What I meant by "error" is "compile-time error". I try to always use "error" that way, to use "warning" to mean a non-error that is enabled by default, and "lint" to mean a non-error that is disabled by default. I use "diagnostic" to mean any of the above.

Somehow I'd gotten the impression that you were proposing that this be an error. I'm not sure how, and you clearly labeled the issue as a "Lint request", so I should have realized that that's not what you had in mind.

> We just need to choose an approach which is in line with existing policies.

It's a bit fuzzy, but we don't allow false positives on lints unless it's not possible to avoid false positives without introducing false negatives AND the risks to a user of having false negatives is high enough to justify the annoyance of false positives. The latter is, of course, purely a judgement call, and not one we always get right.

**Comment by eernstg** [comment]
> What I meant by "error" ... "warning" ... "lint" ...

Very good, we agree completely on that! (Except that I've probably used "warning" less precisely now and then.)

I used the word 'error' a number of times because I was reporting that the tools currently emit some compile-time errors in situations relevant to this issue (especially the CFE).

If it is decided that the CFE should stop reporting these errors (in particular, in response to "this class does not have an implementation of a member which is private to a different library") then I'd very much support the introduction of a lint for the same thing. After all, if my app just crashed with a noSuchMethod exception then I am not going to be a huge lot more happy even if they tell me "but the non-existing thing was a private method!" ;-)

The lint proposed in this issue is somewhat different: It attempts to make a distinction between safe and unsafe invocations of private members. That's a tricky exercise (e.g., because it involves precise knowledge about which classes are "leaking" to other libraries), but it might be possible.
